flytepropeller/pkg/compiler/validators/utils.go

```diff
@@ -170,130 +169,250 @@ func UnionDistinctVariableMaps(m1, m2 map[string]*core.Variable) (map[string]*co
 	return res, nil
 }
 
-func buildMultipleTypeUnion(innerType []*core.LiteralType) *core.LiteralType {
-	var variants []*core.LiteralType
-	isNested := false
+func GetTagForType(x *core.LiteralType) string {
+	if x.GetStructure() == nil {
+		return ""
+	}
+	return x.GetStructure().GetTag()
+}
 
-	for _, x := range innerType {
-		unionType := x.GetCollectionType().GetUnionType()
-		if unionType != nil {
-			isNested = true
-			variants = append(variants, unionType.Variants...)
-		} else {
-			variants = append(variants, x)
+// ValidateLiteralType check if the literal type is valid, return error if the literal is invalid.
+func ValidateLiteralType(lt *core.LiteralType) error {
+	if lt == nil {
+		err := fmt.Errorf("got unknown literal type: [%v].\n"+
+			"Suggested solution: Please update all your Flyte deployment images to the latest version and try again", lt)
+		return err
+	}
+	if lt.GetCollectionType() != nil {
+		return ValidateLiteralType(lt.GetCollectionType())
+	}
+	if lt.GetMapValueType() != nil {
+		return ValidateLiteralType(lt.GetMapValueType())
+	}
+
+	return nil
+}
+
+type instanceChecker interface {
+	isInstance(*core.Literal) bool
+}
+
+type trivialInstanceChecker struct {
+	literalType *core.LiteralType
+}
+
+func (t trivialInstanceChecker) isInstance(lit *core.Literal) bool {
+	if _, ok := lit.GetValue().(*core.Literal_Scalar); !ok {
+		return false
+	}
+	targetType := t.literalType
+	if targetType.GetEnumType() != nil {
+		// If t is an enum, it can be created from a string as Enums as just constrained String aliases
+		if _, ok := lit.GetScalar().GetPrimitive().GetValue().(*core.Primitive_StringValue); ok {
+			return true
 		}
 	}
-	unionLiteralType := &core.LiteralType{
-		Type: &core.LiteralType_UnionType{
-			UnionType: &core.UnionType{
-				Variants: variants,
-			},
-		},
+
+	literalType := literalTypeForScalar(lit.GetScalar())
+	err := ValidateLiteralType(literalType)
+	if err != nil {
+		return false
 	}
+	return AreTypesCastable(literalType, targetType)
+}
 
-	if isNested {
-		return &core.LiteralType{
-			Type: &core.LiteralType_CollectionType{
-				CollectionType: unionLiteralType,
-			},
+type noneInstanceChecker struct{}
+
+func (t noneInstanceChecker) isInstance(lit *core.Literal) bool {
+	if lit == nil {
+		return true
+	}
+	_, ok := lit.GetScalar().GetValue().(*core.Scalar_NoneType)
+	return ok
+}
+
+type collectionInstanceChecker struct {
+	literalType *core.LiteralType
+}
+
+func (t collectionInstanceChecker) isInstance(lit *core.Literal) bool {
+	if _, ok := lit.GetValue().(*core.Literal_Collection); !ok {
+		return false
+	}
+	for _, x := range lit.GetCollection().Literals {
```

Comment:
If we add the tuple literal type but still represent the literal as a collection, how do we check if this collection is a tuple or not

Reply:
The type of Literal for tuple will be Collection, but the LiteralType won't. We'll have to create another new `InstanceChecker` (perhaps `tupleInstanceChecker`) to deal with that.